cpp/src/arrow/util/decimal.cc

```diff
@@ -857,19 +857,31 @@ static const Decimal128 ScaleMultipliers[] = {
 Status Decimal128::Rescale(int32_t original_scale, int32_t new_scale,
                            Decimal128* out) const {
   DCHECK_NE(out, NULLPTR);
-  DCHECK_NE(original_scale, new_scale);
-  const int32_t delta_scale = original_scale - new_scale;
+  DCHECK_NE(original_scale, new_scale) << "original_scale == new_scale";
+
+  const int32_t delta_scale = new_scale - original_scale;
+
+  // Fail if rescaling would truncate
```

Comment:
Is it possible that all truncated digits be zero?

Reply:
Meaning something like `decimal128(10, 2)` is the requested type but the input is something like `1.000` (scale of 3, but all zeros).

Reply:
Yes.